### PR TITLE
[fronius] Fix communication errors by retrying failed http requests

### PR DESCRIPTION
**Pull request by jimtng**
* HTTP Requests to the Fronius controller sometimes time out, causing the thing and bridge to go offline. This PR implements retry with delay to mitigate this issue.
* Synchronize access to the executeUrl command used to access Fronius controller which is used both by the bridge and the things. This avoids simultaneous requests to Fronius especially when multiple bridges are configured.
* Refactor the thinghandler (service) registration to use OpenHAB's built in mechanism. Remove disposed handlers from the bridge so that they are no longer refreshed.
* Trigger a refresh when a thing is added. This avoids a lengthy UNKNOWN state when the refresh interval is long.

### CODE REVIEW

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java [hunk 1]
```diff
@@ -69,29 +71,16 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
-            }
-            ThingHandler handler = bridge.getHandler();
-            if (handler instanceof FroniusBridgeHandler) {
-                this.bridgeHandler = (FroniusBridgeHandler) handler;
-            } else {
-                return null;
-            }
+        // this is important so FroniusBridgeHandler::childHandlerInitialized gets called
+        Bridge bridge = getBridge();
+        if (bridge == null) {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java:77]: Please also check that bridge.getHandler() is not null

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java [hunk 2]
```diff
@@ -160,59 +149,75 @@ protected void updateChannel(String channelId) {
...
     /**
-     * do whatever a thing must do to update the values
+     * Called by the bridge to fetch data and update channels
+     *
+     * @param bridgeConfiguration the connected bridge configuration
+     */
+    public void refresh(FroniusBridgeConfiguration bridgeConfiguration) {
+        try {
+            handleRefresh(bridgeConfiguration);
+            updateStatus(ThingStatus.ONLINE);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java:159]: Maybe set ONLINE only if not already ONLINE

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java [hunk 3]
```diff
@@ -84,38 +87,57 @@ public void initialize() {
...
-                } catch (IOException e) {
-                    logger.debug("Connection Error: {}", e.getMessage());
-                }
-
-                if (!online) {
-                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
-                            "hostname or ip is not reachable");
-                } else {
+                    checkBridgeOnline(config);
                     updateStatus(ThingStatus.ONLINE);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java:126]: Maybe set ONLINE only if not already ONLINE

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/FroniusHttpUtil.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    logger.debug("Failed connecting to {} after {} attempts.", url, attemptCount, lastException);
+                    throw new FroniusCommunicationException("Unable to connect", lastException);
+                }
+
+                logger.debug("HTTP error on attempt #{} {}", attemptCount, url);
+                Thread.sleep(500 * attemptCount);
+                attemptCount++;
+            }
+        } catch (InterruptedException e) {
+            throw new FroniusCommunicationException("Interrupted", e);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/FroniusHttpUtil.java:75]: Please first run: Thread.currentThread().interrupt();

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java (hunk 1 above)
Comment (jimtng) [reply bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java]: Added check for bridge.getHandler()

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java (hunk 2 above)
Comment (jimtng) [reply bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java]: added the check

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java (hunk 3 above)
Comment (jimtng) [reply bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java]: added the check

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java [hunk 5]
```diff
@@ -123,7 +123,9 @@ private void startAutomaticRefresh() {
             Runnable runnable = () -> {
                 try {
                     checkBridgeOnline(config);
-                    updateStatus(ThingStatus.ONLINE);
+                    if (getThing().getStatus() != ThingStatus.OFFLINE) {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java:126]: ONLINE

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java [hunk 6]
```diff
@@ -156,7 +156,9 @@ protected void updateChannel(String channelId) {
     public void refresh(FroniusBridgeConfiguration bridgeConfiguration) {
         try {
             handleRefresh(bridgeConfiguration);
-            updateStatus(ThingStatus.ONLINE);
+            if (getThing().getStatus() != ThingStatus.OFFLINE) {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java:159]: ONLINE

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/FroniusHttpUtil.java (hunk 4 above)
Comment (jimtng) [reply bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/FroniusHttpUtil.java:76]: Done, also in FroniusBaseThingHandler.java

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java (hunk 5 above)
Comment (jimtng) [reply bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBridgeHandler.java]: Oops! Fixed.

File: bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java (hunk 6 above)
Comment (jimtng) [reply bundles/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/internal/handler/FroniusBaseThingHandler.java]: Oops! Fixed.

**Review by lolodomo — APPROVED** [review]
LGTM

**Pull request merged**